stl/inc/random

```diff
@@ -2524,8 +2524,30 @@ private:
 
             return _Res;
         } else if (_Par0._Mean < 1.0) {
-            // events are rare, use Poisson distribution
-            _Res = _Par0._Small(_Eng);
+            // Events are rare, use waiting time method.
+            const _Ty1 _Rand = _NRAND(_Eng, _Ty1);
+            {
+                // The exit condition is -log(1 - _Rand)/t > -log(1-p), which is equivalent to _Rand > 1 - (1-p)^t. If
+                // we have a cheap upper bound for 1-(1-p)^t, we can exit early without having to call log. We use two
+                // such bounds, one that is tight for mean ~0 and another for mean ~1. In the first case, Bernoulli's
+                // inequality gives -1+p*t >= -(1-p)^t, so 1 - (1-p)^t <= p*t = mean. For the other bound, 1-(1-p)^t =
+                // 1-(1-p)(1-mean/t)^(t-1) <= 1-(1-p)(1-1/t)^(t-1) <= 1-(1-p)/e.
+                const _Ty1 _Ub =
+                    std::min(_Par0._Mean, _Ty1{3.678794411714423216e-1} * _Par0._Px + _Ty1{6.32120558828557678e-1});
+                if (_Rand > _Ub)
+                    return 0;
+            }
+
+            _Ty1 _Sum = -_CSTD log(_Ty1{1.0} - _Rand) / _Par0._Tx;
+            if (_Sum > -_Par0._Logp1)
+                return 0;
+            _Res = _Ty{1};
+
+            do {
+                _Sum += -_CSTD log(_Ty1{1.0} - _NRAND(_Eng, _Ty1)) / (_Par0._Tx - _Res);
+                ++_Res;
+            } while (_Sum <= -_Par0._Logp1 && _Res != _Par0._Tx);
+            return (_Sum > -_Par0._Logp1) ? static_cast<_Ty>(_Res - _Ty{1}) : _Par0._Tx;

```

Comment:
Could this control flow be reordered to avoid comparing the final `_Sum` with `-_Par0._Logp1` twice and overcounting `_Res`?